numbat/src/list.rs

```diff
@@ -0,0 +1,297 @@
+//! This module defines a custom kind of list used in the [`numbat::Value::List`].
+//! The specificities of this list are that:
+//! - It's based on a `VecDeque`, which means insertion at the beginning or the end is `O(1)`.
+//! - It's stored behind an `Arc`, which makes cloning the value very cheap.
+//! - It tries to reduce the number of allocations as much as possible, even when shared between multiple values.
+//!   If you are the only owner of the list or while you're not pushing elements in it in, never re-allocate.
+
+use std::{collections::VecDeque, fmt, sync::Arc};
+
+use crate::{value::Value, RuntimeError};
+
+/// Reference counted list / list view
+#[derive(Clone, Eq)]
+pub struct NumbatList<T> {
+    /// The original alloc shared between all values
+    alloc: Arc<VecDeque<T>>,
+    /// The indexes accessible to us. If `None` we own the whole allocation
+    view: Option<(usize, usize)>,
+}
+
+impl<T> Default for NumbatList<T> {
+    fn default() -> Self {
+        Self {
+            alloc: Default::default(),
+            view: Default::default(),
+        }
+    }
+}
```

Comment:
Can this be `derive`d?

Reply:
No because I don't want to impose the `Default` trait to `T` and that makes rust dumb.
It won't implements `Default` on a structure if all of its generic parameters don't implements `Debug` as well 😔